### PR TITLE
Python: change default to `python3`

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/issues/18185
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by domenkozar** [comment]
http://hydra.nixos.org/jobset/nixpkgs/python3 builds python packages only, but I guess we should rebuild the whole nixpkgs? I'll change to that.

**Comment by domenkozar** [comment]
Done. `error: undefined variable ‘pythonPackages’ at /nix/store/2m6xrym9ch7j7j7a0jqsvsazhhp22cnp-git-export/pkgs/development/arduino/ino/default.nix:15:36`

**Comment by FRidh** [comment]
Hydra job: http://hydra.nixos.org/eval/1297631?compare=staging

**Comment by FRidh** [comment]
Failures relative to staging:
http://hydra.nixos.org/eval/1297676?compare=staging#tabs-now-fail

I think I've fixed about half of those by now.

**Comment by FRidh** [comment]
latest evaluation
http://hydra.nixos.org/eval/1297698?compare=staging

@domenkozar could we merge this into staging? There are still some important failures, but as far as I can see no more rebuilds. Then on staging we can fix the final issues that prevent merging into master.

**Comment by domenkozar** [comment]
Two issues merging this:

- it compares to staging evaluation that was mostly canceled, so there might be more failures
- we still have to run all nixos tests before it's merged to staging. It's important to not break everything as that will block others from using staging

**Comment by FRidh** [comment]
> it compares to staging evaluation that was mostly canceled, so there might be more failures

Indeed, there's a bit more failures http://hydra.nixos.org/eval/1297698?compare=1297372

**Comment by FRidh** [comment]
There's also the list of errors to consider. This list will include packages that could not be build because a Python dependency is disabled.
http://hydra.nixos.org/jobset/nixpkgs/python3#tabs-errors

That's a *looooong* list.

**Comment by copumpkin** [comment]
Does the LLVM stuff still work? That's the main thing I care about for Darwin purposes.

**Comment by FRidh** [comment]
I've merged everything except the actual switch to Python 3.

**Comment by FRidh** [comment]
@copumpkin all llvm except for 3.4 were building fine.

**Comment by FRidh** [comment]
Some packages are not available for Python 3. Therefore, any package that depends on this one should use Python 2.

- `gyp` (see also https://codereview.chromium.org/1454433002/)
- `pygame` (although the 1.9.2 alpha with patching could be made to work)
- `pygtk`

**Comment by copumpkin** [comment]
I tried merging this into staging and building the darwin stdenv and I didn't notice any difference during the bootstrap. Is LLVM pulling in python2 explicitly?

**Comment by FRidh** [comment]
@copumpkin I pushed already a bunch of changes (explicitly passing in `python2`) into staging. In adbaed081596919b085d398c9f46e576e2fbc6a9 and 284f22d19e1dd4d09ee90669565cc89b7e6eaeb0 I pass in `python2` for LLVM 3.7. The other versions of LLVM work fine with Python 3.5.

**Comment by FRidh** [comment]
If we now make the change to Python 3, we get these extra failures
http://hydra.nixos.org/eval/1298159?compare=staging#tabs-now-fail
as well as a bunch of extra evaluation errors
http://hydra.nixos.org/jobset/nixpkgs/python3#tabs-errors

Quite some more packages need to be changed to `python2` but it is taking a lot of time. Aside from `virtualbox` I think we have all important packages covered and feel that it is at this point up to the package maintainers to make the final fixes.

**Comment by copumpkin** [comment]
Oh I see, thanks!

**Comment by FRidh** [comment]
My system now runs on this branch, and I haven't encountered any issues so far. Judging from the tests I think there's only one issue remaining, and that would be fixing kde 4. Aside from that there's a bunch of lesser important packages that need to be fixed but I'll leave that for the maintainers.

File: pkgs/development/python-modules/pyside/apiextractor.nix [hunk 1]
```diff
@@ -1,7 +1,7 @@
 { stdenv, fetchurl, cmake, libxml2, libxslt, python, sphinx, qt4 }:
 
 stdenv.mkDerivation {
-  name = "${python.libPrefix}-pyside-apiextractor-0.10.10";
```
Comment (FRidh) [inline pkgs/development/python-modules/pyside/apiextractor.nix]: this should likely be mkPythonDerivation

Comment (FRidh) [inline pkgs/development/python-modules/pyside/apiextractor.nix]: reminder to self: this isn't correct yet.

**Comment by FRidh** [comment]
If we want to have Python 3.5 as default some fixes are going to be needed. If I recall correctly Python 3 builds or packages are not yet deterministic. I won't have much time to work on this or Nixpkgs in general the coming months (final phase of PhD). 

cc release manager @globin

**Comment by globin** [comment]
Do you see this being ready for the branch-off at the end of February?

**Comment by FRidh** [comment]
@globin unless someone else picks it up, no. I won't have any time to work on this. Migrating the final packages to Python 3 (or letting them remain at 2) shouldn't be too much work I think, because of that work has already been done.

**Comment by globin** [comment]
I'll include a link to this PR in the timeline announcement, hoping someone will pick this up.

**Comment by FRidh** [comment]
As was discussed on the mailing list some are against simply staging `python = python3;` and want to replace all occurrences of `python` with `python3`. I'm not interested in doing that work so I'm closing this PR.

**Pull request closed without merging**